sdl-core/src/main/scala/io/smartdatalake/workflow/action/generic/transformer/GenericDfsTransformerDef.scala

```diff
@@ -0,0 +1,151 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package io.smartdatalake.workflow.action.generic.transformer
+
+import io.smartdatalake.config.ParsableFromConfig
+import io.smartdatalake.config.SdlConfigObject.ActionId
+import io.smartdatalake.util.hdfs.PartitionValues
+import io.smartdatalake.util.spark.{DefaultExpressionData, SparkExpressionUtil}
+import io.smartdatalake.workflow.dataframe.GenericDataFrame
+import io.smartdatalake.workflow.dataframe.spark.{SparkDataFrame, SparkSubFeed}
+import io.smartdatalake.workflow.{ActionPipelineContext, DataFrameSubFeed}
+import org.apache.spark.sql.DataFrame
+
+import scala.reflect.runtime.universe
+import scala.reflect.runtime.universe.{Type, typeOf}
+
+
+/**
+ * Interface to implement GenericDataFrame transformers working with many inputs and many outputs (n:m)
+ * Note that this interface cannot be parsed from config, it's only used for programmatically defined transformers.
+ * Check GenericDfsTransformer trait to implement transformers that should be parsed from config.
+ */
+trait GenericDfsTransformerDef extends PartitionValueTransformer {
+  def name: String
+  def description: Option[String]
+
+  /**
+   * Optional function to implement validations in prepare phase.
+   */
+  def prepare(actionId: ActionId)(implicit context: ActionPipelineContext): Unit = Unit
+
+  /**
+   * Function to be implemented to define the transformation between many inputs and many outputs (n:m)
+   * @param actionId id of the action which executes this transformation. This is mainly used to prefix error messages.
+   * @param partitionValues partition values to transform
+   * @param dfs Map of (dataObjectId, DataFrame) tuples available as input
+   * @return Map of transformed (dataObjectId, DataFrame) tuples
+   */
+  def transform(actionId: ActionId, partitionValues: Seq[PartitionValues], dfs: Map[String,GenericDataFrame])(implicit context: ActionPipelineContext): Map[String,GenericDataFrame]
+
+  /**
+   * Declare supported Language for transformation.
+   * Can be DataFrameSubFeed to work with GenericDataFrame, or SparkSubFeed to work with Spark-DataFrames
+   */
+  private[smartdatalake] def getSubFeedSupportedType: Type = typeOf[DataFrameSubFeed]
+
+  private[smartdatalake] def applyTransformation(actionId: ActionId, partitionValues: Seq[PartitionValues], dfs: Map[String,GenericDataFrame])(implicit context: ActionPipelineContext): (Map[String,GenericDataFrame], Seq[PartitionValues]) = {
+    val transformedDfs = transform(actionId, partitionValues, dfs)
+    val transformedPartitionValues = transformPartitionValues(actionId, partitionValues).map(_.values.toSeq.distinct)
+      .getOrElse(partitionValues)
+    (transformedDfs,transformedPartitionValues)
+  }
+}
+
+/**
+ * Interface to implement Spark-DataFrame transformers working with many inputs and many outputs (n:m)
```

Comment:
I guess this comment is meant for SparkDfsTransformer and not GenericDfsTransformer?

Reply:
Done